tests/tests_geomstats/test_special_orthogonal.py

```diff
@@ -1,289 +1,1414 @@
+                + (1 - phi) / 3 * gs.ones([3, 3])
+                + gs.pi / (10 * gs.sqrt(3.0)) * skew
+            )
+            inv_jacobian = gs.linalg.inv(jacobian)
+            aux = gs.dot(inv_jacobian, expected)
+            rot_vec_2 = SpecialOrthogonal(3, "vector").compose(rot_vec_base_point, aux)
+
+            smoke_data = [
+                dict(
+                    point=rot_vec_base_point,
+                    base_point=rot_vec_base_point,
+                    expected=gs.array([0.0, 0.0, 0.0]),
+                ),
+                dict(
+                    point=rot_vec_2,
+                    base_point=rot_vec_base_point,
+                    expected=expected,
+                ),
+            ]
+            return self.generate_tests(smoke_data)
+
+        def distance_broadcast_data(self):
+            smoke_data = [dict(n=2)]
+            return self.generate_tests(smoke_data)
+
+    testing_data = TestDataBiInvariantMetric()
+
+    def test_squared_dist_is_less_than_squared_pi(self, point_1, point_2):
         """
-        This tests that the composition of
-        rotation_vector_from_matrix
-        and
-        matrix_from_rotation_vector
-        is the identity.
+        This test only concerns the canonical metric.
+        For other metrics, the scaling factor can give
+        distances above pi.
         """
-        group = SpecialOrthogonal(n=3)
-        groupvec = SpecialOrthogonal(n=3, point_type="vector")
-        point = groupvec.random_point()
-        rot_mat = group.matrix_from_rotation_vector(point)
-        result = group.rotation_vector_from_matrix(rot_mat)
+        group = SpecialOrthogonal(3, "vector")
+        metric = self.metric(SpecialOrthogonal(3, "vector"))
+        point_1 = group.regularize(point_1)
+        point_2 = group.regularize(point_2)
 
-        expected = self.group.regularize(point)
+        sq_dist = metric.squared_dist(point_1, point_2)
+        diff = sq_dist - gs.pi**2
+        self.assertTrue(
+            diff <= 0 or abs(diff) < EPSILON, "sq_dist = {}".format(sq_dist)
+        )
 
+    def test_exp(self, tangent_vec, base_point, expected):
+        metric = self.metric(SpecialOrthogonal(3, "vector"))
+        result = metric.exp(tangent_vec, base_point)
         self.assertAllClose(result, expected)
+
+    def test_log(self, point, base_point, expected):
+        metric = self.metric(SpecialOrthogonal(3, "vector"))
+        result = metric.log(point, base_point)
+        self.assertAllClose(result, expected)
+
+    @geomstats.tests.np_autograd_and_tf_only
+    def test_distance_broadcast(self, n):
+        group = SpecialOrthogonal(n=n)
+        point = group.random_point(5)
+        result = group.bi_invariant_metric.dist_broadcast(point[:3], point)
+        expected = []
+        for a in point[:3]:
+            expected.append(group.bi_invariant_metric.dist(a, point))
+        expected = gs.stack(expected)
+        self.assertAllClose(result, expected)
```

Comment:
shouldn't expected be passed as argument?

Reply:
I don't think so. This is not some testing function in isolation. It is testing whether function output on vectorized inputs is stack of output of function of single-batch inputs. So no need for expected